Add metrics scope contextmanager

lgeiger opened the pull request:
This allows us to configure training metrics using a context manager similar to `tf.keras.utils.custom_object_scope` or `tf.name_scope`.

I disabled the metrics by default since they can slightly increase runtime and memory consumption.

The API I propose is the following:
```python
with lq.metrics.scope(["flip_ratio"]):
    model = tf.keras.models.Sequential([lq.layers.QuantDense(3, kernel_quantizer="ste_sign", input_shape=(32,))])
model.compile(loss="mse", optimizer="sgd")
```

I still need to add documentation, but I though it would be good to get some feedback first.

koenhelwegen commented:
Awesome! I like this much better. In your example, does the `model.compile()` statement needs to be within the context scope? And the old API is still available right?

lgeiger commented:
> In your example, does the `model.compile()` statement needs to be within the context scope?

Good point. It is not necessary. I updated the unit test and example above.

> And the old API is still available right?

I intend to remove the metrics `kwarg` since it was not documented and only intended as a internal escape hatch if we run into unexpected issues.

MariaHeuss commented:
It looks good but I would not put too much weight on my opinion for this one.

koenhelwegen commented:
I would be in favour of turning the `metrics` kwarg into a permanent, documented feature. It may very well be that a researcher is interested in monitoring only one or two layers, especially as we add more custom metrics. Then a context manager is a bit awkward. This is also what I'm used to from TF (i.e. using a context manager to set a default, and where needed overwriting that with kwargs).

But it adds additional code surface and TF itself is moving away from this coding style a feel, so I'm also fine with not having it.

lgeiger commented:
> I would be in favour of turning the metrics kwarg into a permanent, documented feature. It may very well be that a researcher is interested in monitoring only one or two layers, especially as we add more custom metrics. Then a context manager is a bit awkward. This is also what I'm used to from TF (i.e. using a context manager to set a default, and where needed overwriting that with kwargs).

After some thinking I agree, that this is a useful feature.

I updated the PR and added documentation. This is ready for a thorough review.

koenhelwegen commented on larq/layers.py:
```diff
@@ -75,6 +75,8 @@ class QuantDense(QuantizerBase, tf.keras.layers.Dense):
     kernel_constraint: Constraint function applied to the `kernel` weights matrix.
     bias_constraint: Constraint function applied to the bias vector.
+    metrics: An array of metrics to add to the layer. If `None` the metrics set in
+        `larq.metrics.scope` are used. Available metrics: `flip_ration`.
```
```suggestion
        `larq.metrics.scope` are used. Available metrics: `flip_ratio`.
```

Would it make sense to make this a hyperlink?

lgeiger replied:
Currently we don't have a documentation page for `flip_ratio` since I am still unsure if we want to change the API of the `FlipRatio` class in the future. I think it is fine to keep it as is and we can add more documentation on metrics when we add more.

koenhelwegen commented on larq/layers.py:
```diff
@@ -165,6 +169,8 @@ class QuantConv1D(QuantizerBase, tf.keras.layers.Conv1D):
     kernel_constraint: Constraint function applied to the kernel matrix.
     bias_constraint: Constraint function applied to the bias vector.
+    metrics: An array of metrics to add to the layer. If `None` the metrics set in
+        `larq.metrics.scope` are used. Available metrics: `flip_ration`.
```
```suggestion
        `larq.metrics.scope` are used. Available metrics: `flip_ratio`.
```

lgeiger replied:
Good catch

koenhelwegen commented on larq/layers.py:
```diff
@@ -268,6 +276,8 @@ class QuantConv2D(QuantizerBase, tf.keras.layers.Conv2D):
     kernel_constraint: Constraint function applied to the kernel matrix.
     bias_constraint: Constraint function applied to the bias vector.
+    metrics: An array of metrics to add to the layer. If `None` the metrics set in
+        `larq.metrics.scope` are used. Available metrics: `flip_ration`.
```
```suggestion
        `larq.metrics.scope` are used. Available metrics: `flip_ratio`.
```

koenhelwegen commented on larq/layers.py:
```diff
@@ -378,6 +390,8 @@ class QuantConv3D(QuantizerBase, tf.keras.layers.Conv3D):
     kernel_constraint: Constraint function applied to the kernel matrix.
     bias_constraint: Constraint function applied to the bias vector.
+    metrics: An array of metrics to add to the layer. If `None` the metrics set in
+        `larq.metrics.scope` are used. Available metrics: `flip_ration`.
```
```suggestion
        `larq.metrics.scope` are used. Available metrics: `flip_ratio`.
```

koenhelwegen replied:
I will leave the rest for a search replace ;)

koenhelwegen commented on larq/layers_test.py:
```diff
@@ -77,16 +77,17 @@ def test_binarization(self, quantized_layer, layer, input_shape, kwargs):
-            ),
-            input_data=input_data,
-        )
+        with lq.metrics.scope(["flip_ratio"]):
```
I do think we should have unit test for the `metrics` kwarg as well

lgeiger replied:
I added a unit test in the latest commit
